File: wire-android-automation/common/src/main/java/com/wearezeta/auto/common/email/handlers/InbucketChangesListener.java
package com.wearezeta.auto.common.email.handlers;

import com.wearezeta.auto.common.email.inbucket.models.Message;
import com.wearezeta.auto.common.misc.Timedelta;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.time.Instant;
import java.util.List;
import java.util.Map;

public class InbucketChangesListener extends AbstractMailboxChangesListener {

    private final InbucketMailbox mailbox;
    // Example date: "2022-07-13T14:59:38.314900766Z"
    private final DateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSSSSSSS'Z'");

    public InbucketChangesListener(InbucketMailbox mailbox, Map<String, String> expectedHeaders,
                                   Timedelta timeout, Timedelta rejectMessagesBefore) {
        super(mailbox, expectedHeaders, timeout, rejectMessagesBefore);
        this.mailbox = mailbox;
    }

    @Override
    public String call() throws Exception {
        final String deliveredTo = mailbox.getEmailAddress();
        final Timedelta started = Timedelta.now();
        do {
            final List<Message> messages = this.mailbox.getRecentMessages(deliveredTo);
            log.info(String.format("Got %s incoming message(s) for %s", messages.size(), deliveredTo));
            for (Message message: messages)  {
                if (this.areAllHeadersInMessage(message)) {
                    Instant messageDate = Instant.parse(message.getDate());
                    Instant listeningStartDate = Instant.ofEpochSecond(rejectMessagesBefore.asSeconds());
                    if (messageDate.isAfter(listeningStartDate)) {
                        log.info(String.format(
                                "Message accepted because message date (%s) is after start of listening (%s)",
                                messageDate,
                                listeningStartDate));
                        return this.mailbox.getMessageSource(deliveredTo, message.getId());
                    }
                    log.severe(String.format(
                            "Message rejected because message date (%s) is before start of listening (%s)",
                            messageDate,
                            listeningStartDate));
                }
            }
            Timedelta.ofSeconds(2).sleep();
        } while (Timedelta.now().isDiffLessOrEqual(started, timeout));
        throw new RuntimeException(String.format("Email message containing headers %s has not been found within %s",
                this.expectedHeaders.toString(), this.timeout));
    }

    protected boolean areAllHeadersInMessage(Message message) {
        for (Map.Entry<String, String> expectedHeader : this.expectedHeaders.entrySet()) {
            boolean isHeaderFound = false;
            final String expectedHeaderName = expectedHeader.getKey();
            final String expectedHeaderValue = expectedHeader.getValue();
                List<String> headerValues = message.getHeader().get(expectedHeaderName);
                log.info(String.format("Checking if the email message contains %s: %s header",
                        expectedHeaderName, expectedHeaderValue));
                if (headerValues != null) {
                    for (String headerValue : headerValues) {
                        log.info(String.format("%s: %s -> %s", expectedHeaderName, headerValue, expectedHeaderValue));
                        if (headerValue.equals(expectedHeaderValue)) {
                            log.info(String.format("The expected header value '%s' is found in the email",
                                    expectedHeaderValue));
                            isHeaderFound = true;
                            break;
                        }
                    }
                }
                log.info(String.format("Header %s with value %s found: %s", expectedHeaderName, expectedHeaderValue,
                        isHeaderFound));
            if (!isHeaderFound) {
                return false;
            }
        }
        return true;
    }

}

File: tests/testsCore/src/main/kotlin/com/wire/android/tests/core/email/InbucketChangesListener.kt
/*
 * Wire
 * Copyright (C) 2025 Wire Swiss GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see http://www.gnu.org/licenses/.
 */
package com.wire.android.tests.core.email

import com.wire.android.tests.core.models.Message
import com.wire.android.tests.core.utils.Timedelta
import java.text.DateFormat
import java.text.SimpleDateFormat
import java.time.Instant
import java.util.*

class InbucketChangesListener(
    private val mailbox: InbucketMailbox,
    expectedHeaders: Map<String, String>,
    timeout: Timedelta,
    rejectMessagesBefore: Timedelta
) : AbstractMailboxChangesListener(mailbox, expectedHeaders, timeout, rejectMessagesBefore) {

    // Example date: "2022-07-13T14:59:38.314900766Z"
    private val dateFormat: DateFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSSSSSSS'Z'")

    @Throws(Exception::class)
    override fun call(): String {
        val deliveredTo = mailbox.emailAddress
        val started = Timedelta.now()

        do {
            val messages = mailbox.getRecentMessages(deliveredTo)
            log.info("Got ${messages.size} incoming message(s) for $deliveredTo")

            for (message in messages) {
                if (areAllHeadersInMessage(message)) {
                    val messageDate = Instant.parse(message.date)
                    val listeningStartDate = Instant.ofEpochSecond(rejectMessagesBefore.asSeconds().toLong())

                    if (messageDate.isAfter(listeningStartDate)) {
                        log.info("Message accepted because message date ($messageDate) is after start of listening ($listeningStartDate)")
                        return mailbox.getMessageSource(deliveredTo, message.id)
                    }
                    log.severe("Message rejected because message date ($messageDate) is before start of listening ($listeningStartDate)")
                }
            }
            Timedelta.ofSeconds(2.0).sleep()
        } while (Timedelta.now().isDiffLessOrEqual(started, timeout))

        throw RuntimeException("Email message containing headers ${expectedHeaders} has not been found within $timeout")
    }

    protected fun areAllHeadersInMessage(message: Message): Boolean {
        for ((expectedHeaderName, expectedHeaderValue) in expectedHeaders) {
            var isHeaderFound = false
            log.info("Checking if the email message contains $expectedHeaderName: $expectedHeaderValue header")

            val headerValues = message.header?.get(expectedHeaderName)
            if (headerValues != null) {
                for (headerValue in headerValues) {
                    log.info("$expectedHeaderName: $headerValue -> $expectedHeaderValue")
                    if (headerValue == expectedHeaderValue) {
                        log.info("The expected header value '$expectedHeaderValue' is found in the email")
                        isHeaderFound = true
                        break
                    }
                }
            }
            log.info("Header $expectedHeaderName with value $expectedHeaderValue found: $isHeaderFound")
            if (!isHeaderFound) return false
        }
        return true
    }
}
